### PR TITLE
**Breaking change for types** plus more

### PR DESCRIPTION
**Pull request by MichaelS11**
**Breaking change for types**

Make arrays last two args now optional
Changed type to be an expression
Type expression converts to string for type lookup
DefineType is now local scope
Added DefineReflectType
Added DefineGlobalType & DefineGlobalReflectType
Added make multidimensional array support
Improved multidimensional array support
Added AddPackage

### CODE REVIEW

**Comment by MichaelS11** [comment]
Packages moving out of core coming soon now that types are local scope.

File: parser/parser.go.y [hunk 1]
```diff
@@ -249,6 +250,16 @@ stmt_default :
 		$$ = &ast.DefaultStmt{Stmts: $4}
 	}
 
+array_count :
+	array_count '[' ']'
+	{
+		$$ = ast.ArrayCount{Count: $1.Count + 1}
```
Comment (MichaelS11) [inline parser/parser.go.y]: Is there a way to do $1.Count++ ? I was having issues getting that to work.

**Comment by mattn** [comment]
Hmm, it is impossible to change string literals in new/make?

**Comment by MichaelS11** [comment]
My first reaction was also to keep literals for new/make, but then I started thinking about it more. Having expressions for new/make can be really useful and powerful with very little overhead. A simple example:

```
...
a = make([]typeOf(b))
...
```

If you did that with literals it would have to be a case statement, maybe a really large case statement depending on how many types.

**Comment by MichaelS11** [comment]
Looking into creating a more general SortArray for core

**Comment by MichaelS11** [comment]
Working on refactor for function code to see about handling sort functions

**Comment by MichaelS11** [comment]
This should be ready. Now can do sort functions inside of VM.

**Comment by mattn** [comment]
Well, is not possible to implement non-quoted like `make([]bool)` ?

**Comment by MichaelS11** [comment]
Yes it is possible, but probably not good to have the string literals syntax as well the expression syntax . Probably should choice one or the other. I personally like the expression syntax.

**Comment by MichaelS11** [comment]
Is sounds like you want me to change it back to string literals syntax. Is that what you would like me to do?

**Comment by mattn** [comment]
>Is sounds like you want me to change it back to string literals syntax. Is that what you would like me to do?

Yes, I want to keep compatibility for Go as possible.

**Comment by MichaelS11** [comment]
Changed to string literal

**Comment by MichaelS11** [comment]
What do you think?

**Comment by MichaelS11** [comment]
This good to go?

**Comment by mattn** [comment]
Well, as I mentioned in above, I hope to keep implementation of `make([]string 3)`.

**Comment by MichaelS11** [comment]
One can do any of the following:
```
make([]string)
make([]string, 3)
make([]string, 1, 3)
make([][]string)
make([][]string, 3)
make([][]string, 1, 3)
```
etc...

It can not do `make([]string 3)` without comma.

**Comment by mattn** [comment]
comma is my mistake. but it seems that some string literals are remaining in example?

**Comment by mattn** [comment]
Big Thanks.

**Pull request merged**